database/sqlite3.ksy

```diff
@@ -32,213 +32,431 @@ doc: |
 seq:
-  - id: magic
-    contents: ["SQLite format 3", 0]
-  - id: len_page_mod
-    type: u2
-    doc: |
-      The database page size in bytes. Must be a power of two between
-      512 and 32768 inclusive, or the value 1 representing a page size
-      of 65536.
-  - id: write_version
-    type: u1
-    enum: versions
-  - id: read_version
-    type: u1
-    enum: versions
-  - id: reserved_space
-    type: u1
-    doc: Bytes of unused "reserved" space at the end of each page. Usually 0.
-  - id: max_payload_frac
-    type: u1
-    doc: Maximum embedded payload fraction. Must be 64.
-  - id: min_payload_frac
-    type: u1
-    doc: Minimum embedded payload fraction. Must be 32.
-  - id: leaf_payload_frac
-    type: u1
-    doc: Leaf payload fraction. Must be 32.
-  - id: file_change_counter
-    type: u4
-  - id: num_pages
-    type: u4
-    doc: Size of the database file in pages. The "in-header database size".
-  - id: first_freelist_trunk_page
-    type: u4
-    doc: Page number of the first freelist trunk page.
-  - id: num_freelist_pages
-    type: u4
-    doc: Total number of freelist pages.
-  - id: schema_cookie
-    type: u4
-  - id: schema_format
-    type: u4
-    doc: The schema format number. Supported schema formats are 1, 2, 3, and 4.
-  - id: def_page_cache_size
-    type: u4
-    doc: Default page cache size.
-  - id: largest_root_page
-    type: u4
-    doc: The page number of the largest root b-tree page when in auto-vacuum or incremental-vacuum modes, or zero otherwise.
-  - id: text_encoding
-    type: u4
-    enum: encodings
-    doc: The database text encoding. A value of 1 means UTF-8. A value of 2 means UTF-16le. A value of 3 means UTF-16be.
-  - id: user_version
-    type: u4
-    doc: The "user version" as read and set by the user_version pragma.
-  - id: is_incremental_vacuum
-    type: u4
-    doc: True (non-zero) for incremental-vacuum mode. False (zero) otherwise.
-  - id: application_id
-    type: u4
-    doc: The "Application ID" set by PRAGMA application_id.
-  - id: reserved
-    size: 20
-  - id: version_valid_for
-    type: u4
-  - id: sqlite_version_number
-    type: u4
-  - id: root_page
-    type: btree_page
+  - id: header
+    type: database_header
 instances:
-  len_page:
-    value: 'len_page_mod == 1 ? 0x10000 : len_page_mod'
+  pages:
+    type:
+      switch-on: '(_index == header.lock_byte_page_index ? 0 : _index >= header.first_ptrmap_page_index and _index <= header.last_ptrmap_page_index ? 1 : 2)'
+      cases:
+        0: lock_byte_page(_index + 1)
+        1: ptrmap_page(_index + 1)
+        # TODO: Free pages and cell overflow pages are incorrectly interpreted as btree pages
+        # This is unfortunate, but unavoidable since there's no way to recognize these types at
+        # this point in the parser.
+        2: btree_page(_index + 1)
+    pos: 0
+    size: header.page_size
+    repeat: expr
+    repeat-expr: header.page_count
 types:
+  database_header:
```

Comment:
any ideas where these come from? (python)

```
>>> db.header.j
819
>>> db.header.m
489
```

Reply:
Thos somewhat cryptic single letter variables are all references to the documentation.
`J` is defined at https://www.sqlite.org/fileformat2.html#pointer_map_or_ptrmap_pages

> In a database with ptrmap pages, the first ptrmap page is page 2. A ptrmap page consists of an array of 5-byte entries. Let J be the number of 5-byte entries that will fit in the usable space of a page. (In other words, J=U/5.) The first ptrmap page will contain back pointer information for pages 3 through J+2, inclusive. The second pointer map page will be on page J+3 and that ptrmap page will provide back pointer information for pages J+4 through 2*J+3 inclusive. And so forth for the entire database file.

`M` is described in https://www.sqlite.org/fileformat2.html#b_tree_pages

> The amount of payload that spills onto overflow pages also depends on the page type. For the following computations, let U be the usable size of a database page, the total page size less the reserved space at the end of each page. And let P be the payload size. In the following, symbol X represents the maximum amount of payload that can be stored directly on the b-tree page without spilling onto an overflow page and symbol M represents the minimum amount of payload that must be stored on the btree page before spilling is allowed.

The paragraphs after that describe how to calculate the values.

Reply:
> In a database with ptrmap pages, the first ptrmap page is page 2. A ptrmap page consists of an array of 5-byte entries. Let J be the number of 5-byte entries that will fit in the usable space of a page.

J = ptrmap_entry_count_max

> symbol X represents the maximum amount of payload that can be stored directly on the b-tree page without spilling onto an overflow page

> symbol M represents the minimum amount of payload that must be stored on the btree page before spilling is allowed.

X = overflow_payload_size_max
M = overflow_payload_size_min

Reply:
I agree with @milahu - the single-letter "names" are incomprehensible, and it doesn't matter whether any credible source uses them or not. When using a `.ksy` specification, there should be no need to have the original SQLite specification on another monitor to decrypt the nonsensical names.

@pepijnve You can use the `-orig-id` key to keep a reference to the original identifiers if you want, but the main `id` names should be as clear and consistent as possible. See https://doc.kaitai.io/ksy_style_guide.html#transcribing:

> Use `-orig-id` key to specify original names of fields for purposes of maintaining a reference link to parts of original spec, but, otherwise, feel free to use a more consistent and language-neutral approach in naming attributes and types.